rust/src/detect/datasets.rs

```diff
@@ -0,0 +1,307 @@
+        -> i32;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ParseDatasets(
+    set: &Dataset, name: *const c_char, fname: *const c_char, fmode: *const c_char,
+    dstype: DatasetType,
+) -> i32 {
+    let file_string = unwrap_or_return!(CStr::from_ptr(fname).to_str(), -2);
+    let mode = unwrap_or_return!(CStr::from_ptr(fmode).to_str(), -2);
+    let set_name = unwrap_or_return!(CStr::from_ptr(name).to_str(), -2);
+    let filename = Path::new(file_string);
+    let mut is_dataset = false;
+    let mut is_datarep = false;
+    if let Ok(lines) = read_lines(filename, mode) {
+        for line in lines.map_while(Result::ok) {
+            let v: Vec<&str> = line.split(',').collect();
+            // Ignore empty and invalid lines in dataset/rep file
+            if v.is_empty() || v.len() > 2 {
+                continue;
+            }
+
+            if v.len() == 1 {
+                if is_datarep {
+                    SCLogError!(
+                        "Cannot mix dataset and datarep values for set {} in {}",
+                        set_name,
+                        filename.display()
+                    );
+                    return -2;
+                }
+                // Dataset
+                is_dataset = true;
+            } else {
+                if is_dataset {
+                    SCLogError!(
+                        "Cannot mix dataset and datarep values for set {} in {}",
+                        set_name,
+                        filename.display()
+                    );
+                    return -2;
+                }
+                // Datarep
+                is_datarep = true;
+            }
+            match dstype {
+                DatasetType::DSString => {
+                    if process_string_set(set, v, set_name, filename, is_dataset) == -1 {
+                        continue;
+                    }
+                }
+                DatasetType::DSMd5 => {
+                    if process_md5_set(set, v, set_name, filename, is_dataset) == -1 {
+                        continue;
+                    }
+                }
+                DatasetType::DSSha256 => {
+                    if process_sha256_set(set, v, set_name, filename, is_dataset) == -1 {
+                        continue;
+                    }
+                }
+                DatasetType::DSIpv4 => {
+                    if process_ipv4_set(set, v, set_name, filename, is_dataset) == -1 {
+                        continue;
+                    }
+                }
+                DatasetType::DSIpv6 => {
+                    if process_ipv6_set(set, v, set_name, filename, is_dataset) == -1 {
+                        continue;
+                    }
+                }
+            }
+        }
+    } else {
+        return -1;
+    }
+
+    0
+}
+
+unsafe fn process_string_set(
+    set: &Dataset, v: Vec<&str>, set_name: &str, filename: &Path, is_dataset: bool,
```

Comment:
should be `is_dataset` actually be something like `with_rep`? Both are datasets, just one is with rep the other w/o.

Reply:
indeed that'd be better and correct